caikit/core/modules/saver.py

```diff
@@ -0,0 +1,334 @@
+from .loader import ModuleLoader
+from caikit.config import get_config
+from caikit.core.modules.config import ModuleConfig
+from caikit.core.toolkit import ObjectSerializer
+from caikit.core.toolkit.errors import error_handler
+from caikit.core.toolkit.wip_decorator import TempDisableWIP
+
+log = alog.use_channel("MODULE")
+error = error_handler.get(log)
+
+
+class ModuleSaver:
+    """A module saver that provides common functionality used for saving modules and also a context
+    manager that cleans up gracefully in case an error is encountered during the save process.
+    """
+
+    SAVED_KEY_NAME = "saved"
+    CREATED_KEY_NAME = "created"
+    TRACKING_KEY_NAME = "tracking_id"
+    MODULE_VERSION_KEY_NAME = "version"
+    MODULE_ID_KEY_NAME = "module_id"
+    MODULE_CLASS_KEY_NAME = "module_class"
+
+    def __init__(self, module: ModuleBase, model_path):
+        """Construct a new module saver.
+
+        Args:
+            module:  caikit.core.module.Module
+                The instance of the module to be saved.
+            model_path:  str
+                The absolute path to the directory where the model will be saved.  If this directory
+                does not exist, it will be created.
+        """
+        self.model_path = os.path.normpath(model_path)
+
+        # Get possibly nested caikit library path
+        module_path = module.__module__
+        lib_name_generator = (
+            k
+            for k, v in get_config().libraries.items()
+            if module_path.startswith(v.module_path)
+        )
+        try:
+            self.library_name = next(lib_name_generator)
+        except StopIteration:
+            # This assumes no nested module path by default
+            self.library_name = module_path.split(".")[0]  # tests
+
+        try:
+            self.library_version = metadata.version(self.library_name)
+        except metadata.PackageNotFoundError:
+            log.debug("<COR25991305D>", "No library version found")
+            if (
+                self.library_name in get_config().libraries
+                and "version" in get_config().libraries[self.library_name]
+            ):
+                self.library_version = get_config().libraries[self.library_name].version
+            else:
+                self.library_version = "0.0.0"
+
+        self.config = {
+            self.library_name + "_version": self.library_version,
+            self.CREATED_KEY_NAME: str(datetime.datetime.now()),
+            self.SAVED_KEY_NAME: str(datetime.datetime.now()),
+            "name": module.MODULE_NAME,
+            self.TRACKING_KEY_NAME: str(uuid.uuid4()),
+            self.MODULE_ID_KEY_NAME: module.MODULE_ID,
+            self.MODULE_CLASS_KEY_NAME: module.MODULE_CLASS,
+            self.MODULE_VERSION_KEY_NAME: module.MODULE_VERSION,
+        }
+
+        # Temp disable wip for following invocation to not log warnings for downstream
+        # usage of ModuleSaver
+        with TempDisableWIP():
```

Comment:
(unrelated to this PR) Should we move the `metadata` stuff out of WIP? Seems hacky to have it in WIP, then temporarily disable it

Reply:
Yeah, I think that'd be fine. It's been around for quite a while and nobody has complained too much about it at least!